### PR TITLE
 feat(gatsby): do not store result in variable

### PR DESCRIPTION
**Pull request by wardpeet**
## Description

We shouldn't store writePage data in a result when we don't need it, it keeps it in memory too long.

This dropped memory usage from 8GB to 4GB in our customer.

## Related Issues

Fixes [ch26306]

### CODE REVIEW

File: packages/gatsby/src/utils/webpack-status.ts [hunk 1]
```diff
@@ -10,9 +10,10 @@ export function markWebpackStatusAsPending(): void {
   isPendingStatus = true
 }
 
-export function markWebpackStatusAsDone(): void {
+export async function markWebpackStatusAsDone(): Promise<void> {
   isPendingStatus = false
+
   if (isFlushEnqueued()) {
-    flush()
+    await flush()
```
Comment (wardpeet) [inline packages/gatsby/src/utils/webpack-status.ts:17]: Adding await wasn't necessary to fix the bug but this seems like a possible timing issue that can occur. Maybe we don't care about awaiting here but than I would like to add a different function that flushes if necessary.

During `gatsby build` we never wait until this possible flush is done. I can't think of when we ever have to flush but this seems like a code smell to me.

**Review by wardpeet — COMMENTED** [review]
When looking at this code snippet:
https://github.com/gatsbyjs/gatsby/pull/30440/files#diff-d61539fe69ce8a4900be5d6476c6bfa63d4b18375d9859c099098e084cfd1841R123-R126

I wonder if we should return the promise of the queue here so we're actually awaiting the current flush

**Comment by KyleAMathews** [comment]
Huh odd. This would certainly work but I'm curious why fastq isn't working as expected. I'll file a bug there tomorrow.

**Comment by wardpeet** [comment]
I messed up a few things in my testing. The drain is actually not reliable when using fastq.promise, not in the callback scenario. When using callbacks it felt like things were called synchronously instead of async. As the queue was always done when idle was called for.

By adding an extra tick, node has a moment to spare and could run gc and my problems went away

File: packages/gatsby/src/utils/page-data.ts [hunk 2]
```diff
@@ -204,11 +204,15 @@ export async function flush(): Promise<void> {
...
 
   if (!flushQueue.idle()) {
-    await new Promise(resolve => {
-      flushQueue.drain = resolve as () => unknown
+    await new Promise<void>(resolve => {
+      flushQueue.drain = resolve
     })
   }
 
+  await new Promise<void>(resolve => {
```
Comment (KyleAMathews) [inline packages/gatsby/src/utils/page-data.ts:212]: Confused about the relationship between this & memory usage. Also how the queue would empty almost immediately in your testing? The work of the queue is to write out files which is pretty fast but still should take a number of seconds for 10s of thousands of paths it sounds like your testing.

**Comment by wardpeet** [comment]
This seems to fix the problem now but the underlying one is that page-data writes take up to much memory, going to close it for now.

**Pull request closed without merging**